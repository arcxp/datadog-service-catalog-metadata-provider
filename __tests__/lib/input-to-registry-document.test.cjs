/**
 * This test covers schema validation for all v2 schema.
 * All v3 schema validation will take place in separate tests.
 * @jest-environment node
 * @group ci
 * @author Mike Stemle <hello@mikestemle.com>
 **/

const YAML = require('yaml')
const core = require('@actions/core')

const Ajv = require('ajv')
const ddSchema_v2 = require('../data/datadog-service-catalog-schema-v2.json')
const ddSchema_v2_1 = require('../data/datadog-service-catalog-schema-v2.1.json')
const ddSchema_v2_2 = require('../data/datadog-service-catalog-schema-v2.2.json')
const validate_v2 = new Ajv({ strict: false, validateFormats: false }).compile(
  ddSchema_v2
)
const validate_v2_1 = new Ajv({
  strict: false,
  validateFormats: false,
}).compile(ddSchema_v2_1)
const validate_v2_2 = new Ajv({
  strict: false,
  validateFormats: false,
}).compile(ddSchema_v2_2)

const {
  inputsToRegistryDocument,
} = require('../../lib/input-to-registry-document')

/// ----------------------------
/// SCHEMA V2.1
/// ----------------------------

describe('input-to-registry-document.js - schema v2', () => {
  test('#inputsToRegistryDocument() - 100% input', async () => {
    const testInput = `
---
schema-version: v2
datadog-key: FAKE_KEY
datadog-app-key: FAKE_KEY
service-name: datadog-service-catalog-metadata-provider-test
team: Team Name Here
email: 'team-name-here@fakeemaildomainthatdoesntexist.com'
slack-support-channel: 'https://fakeorg.slack.com/archives/A0000000000'
repo: https://github.com/arcxp/datadog-service-catalog-metadata-provider
tags: |
  - 'application:GitHub Action Config Test'
  - env:prod
  - infrastructure:serverless
  - language:nodejs
  - other :   value
repos: |
  - url: https://github.com/actions/toolkit
    provider: Github
    name: "@actions/toolkit"
links: |
  - name: AMI Version Status Dashboard
    url: https://thisisanentirelyfakeurl.seriouslythisisafakehostname.com/dashboard
    type: dashboard
docs: |
  - name: GitHub Actions!
    url: https://github.com/features/actions
    provider: Github
integrations: |
  opsgenie:
    service-url: https://yourorghere.app.opsgenie.com/service/00000000-0000-0000-0000-000000000000
    region: US
  pagerduty: https://my-org.pagerduty.com/service-directory/PMyService
contacts: |
  - name: DBA Team Email Alias
    type: email
    contact: dba-team-name-here@fakeemaildomainthatdoesntexist.com
    `

    core.__setInputsObject(YAML.parse(testInput))
    const inputs = inputsToRegistryDocument()
    expect(validate_v2(inputs)).toStrictEqual(true)
    if (validate_v2.errors) {
      console.error(validate_v2.errors)
    }
    expect(validate_v2.errors).toBeNull()
    expect(inputs).toMatchSnapshot()
    expect(Array.isArray(inputs.links)).toBeTruthy()
    expect(inputs.links.length).toEqual(1)
  })

  test('#inputsToRegistryDocument() - normal input', async () => {
    const testInput = `
---
datadog-key: FAKE_KEY
datadog-app-key: FAKE_KEY
service-name: datadog-service-catalog-metadata-provider-test
team: Team Name Here
email: 'team-name-here@fakeemaildomainthatdoesntexist.com'
slack-support-channel: 'https://fakeorg.slack.com/archives/A0000000000'
repo: https://github.com/arcxp/datadog-service-catalog-metadata-provider
tags: |
  - 'application:GitHub Action Config Test'
  - env:prod
  - infrastructure:serverless
  - language:nodejs
  - other :   value
repos: |
  - url: https://github.com/actions/toolkit
    provider: Github
    name: "@actions/toolkit"
links: |
  - name: AMI Version Status Dashboard
    url: https://thisisanentirelyfakeurl.seriouslythisisafakehostname.com/dashboard
    type: dashboard
docs: |
  - name: GitHub Actions!
    url: https://github.com/features/actions
    provider: Github
integrations: |
  opsgenie:
    service-url: https://yourorghere.app.opsgenie.com/service/00000000-0000-0000-0000-000000000000
    region: US
    `

    core.__setInputsObject(YAML.parse(testInput))
    const inputs = inputsToRegistryDocument()
    expect(validate_v2(inputs)).toStrictEqual(true)
    if (validate_v2.errors) {
      console.error(validate_v2.errors)
    }
    expect(validate_v2.errors).toBeNull()
    expect(inputs).toMatchSnapshot()
  })

  test('#inputsToRegistryDocument() - less input', async () => {
    const testInput = `
---
schema-version: v2
datadog-key: FAKE_KEY
datadog-app-key: FAKE_KEY
service-name: datadog-service-catalog-metadata-provider-test
team: Team Name Here
email: 'team-name-here@fakeemaildomainthatdoesntexist.com'
slack-support-channel: 'https://fakeorg.slack.com/archives/A0000000000'
repo: https://github.com/arcxp/datadog-service-catalog-metadata-provider
tags: |
  - 'application:GitHub Action Config Test'
  - env:prod
  - infrastructure:serverless
  - language:nodejs
integrations: |
  opsgenie:
    service-url: https://yourorghere.app.opsgenie.com/service/00000000-0000-0000-0000-000000000000
    region: US
    `

    core.__setInputsObject(YAML.parse(testInput))
    const inputs = inputsToRegistryDocument()
    expect(validate_v2(inputs)).toStrictEqual(true)
    if (!validate_v2(inputs)) {
      console.error(validate_v2_1.errors)
    }
    expect(validate_v2.errors).toBeNull()
    expect(inputs).toMatchSnapshot()
    expect(inputs.links).toBeUndefined()
    expect(inputs.docs).toBeUndefined()
  })
})

/// ----------------------------
/// SCHEMA V2.1
/// ----------------------------

describe('input-to-registry-document.js - schema v2.1', () => {
  test('#inputsToRegistryDocument() - 100% input', async () => {
    const testInput = `
---
schema-version: v2.1
datadog-key: FAKE_KEY
datadog-app-key: FAKE_KEY
service-name: datadog-service-catalog-metadata-provider-test
team: Team Name Here
email: 'team-name-here@fakeemaildomainthatdoesntexist.com'
slack-support-channel: 'https://fakeorg.slack.com/archives/A0000000000'
repo: https://github.com/arcxp/datadog-service-catalog-metadata-provider
tags: |
  - 'application:GitHub Action Config Test'
  - env:prod
  - infrastructure:serverless
  - language:nodejs
  - other :   value
links: |
  - name: AMI Version Status Dashboard
    url: https://thisisanentirelyfakeurl.seriouslythisisafakehostname.com/dashboard
    type: dashboard
  - url: https://github.com/actions/toolkit
    type: repo
    name: "@actions/toolkit"
  - name: GitHub Actions!
    url: https://github.com/features/actions
    type: doc
integrations: |
  opsgenie:
    service-url: https://yourorghere.app.opsgenie.com/service/00000000-0000-0000-0000-000000000000
    region: US
  pagerduty:
    service-url: https://my-org.pagerduty.com/service-directory/PMyService
contacts: |
  - name: DBA Team Email Alias
    type: email
    contact: dba-team-name-here@fakeemaildomainthatdoesntexist.com
    `

    core.__setInputsObject(YAML.parse(testInput))
    const inputs = inputsToRegistryDocument()
    if (!validate_v2_1(inputs)) {
      console.error(validate_v2_1.errors)
    }
    expect(validate_v2_1(inputs)).toStrictEqual(true)
    expect(validate_v2_1.errors).toBeNull()
    expect(inputs).toMatchSnapshot()
    expect(Array.isArray(inputs.links)).toBeTruthy()
    expect(inputs.links.length).toEqual(4)
  })

  test('#inputsToRegistryDocument() - normal input', async () => {
    const testInput = `
---
schema-version: v2.1
datadog-key: FAKE_KEY
datadog-app-key: FAKE_KEY
service-name: datadog-service-catalog-metadata-provider-test
team: Team Name Here
description: This description describes a service.
tier: Tier 1
lifecycle: production
email: 'team-name-here@fakeemaildomainthatdoesntexist.com'
slack-support-channel: 'https://fakeorg.slack.com/archives/A0000000000'
repo: https://github.com/arcxp/datadog-service-catalog-metadata-provider
tags: |
  - 'application:GitHub Action Config Test'
  - env:prod
  - infrastructure:serverless
  - language:nodejs
  - other :   value
links: |
  - url: https://github.com/actions/toolkit
    type: repo
    name: "@actions/toolkit"
    provider: GitHub
  - name: AMI Version Status Dashboard
    url: https://thisisanentirelyfakeurl.seriouslythisisafakehostname.com/dashboard
    type: dashboard
  - name: GitHub Actions!
    url: https://github.com/features/actions
    type: doc
integrations: |
  opsgenie:
    service-url: https://yourorghere.app.opsgenie.com/service/00000000-0000-0000-0000-000000000000
    region: US
    `

    core.__setInputsObject(YAML.parse(testInput))
    const inputs = inputsToRegistryDocument()
    if (!validate_v2_1(inputs)) {
      console.error(validate_v2_1.errors)
    }
    expect(validate_v2_1(inputs)).toStrictEqual(true)
    expect(validate_v2_1.errors).toBeNull()
    expect(inputs).toMatchSnapshot()
  })

  test('#inputsToRegistryDocument() - less input', async () => {
    const testInput = `
---
schema-version: v2.1
datadog-key: FAKE_KEY
datadog-app-key: FAKE_KEY
service-name: datadog-service-catalog-metadata-provider-test
team: Team Name Here
email: 'team-name-here@fakeemaildomainthatdoesntexist.com'
slack-support-channel: 'https://fakeorg.slack.com/archives/A0000000000'
repo: https://github.com/arcxp/datadog-service-catalog-metadata-provider
tags: |
  - 'application:GitHub Action Config Test'
  - env:prod
  - infrastructure:serverless
  - language:nodejs
integrations: |
  opsgenie:
    service-url: https://yourorghere.app.opsgenie.com/service/00000000-0000-0000-0000-000000000000
    region: US
    `

    core.__setInputsObject(YAML.parse(testInput))
    const inputs = inputsToRegistryDocument()
    if (!validate_v2_1(inputs)) {
      console.error(validate_v2_1.errors)
    }
    expect(validate_v2_1(inputs)).toStrictEqual(true)
    expect(validate_v2_1.errors).toBeNull()
    expect(inputs).toMatchSnapshot()
    expect(Array.isArray(inputs.links)).toBeTruthy()
    expect(inputs.links.length).toBe(1)
    expect(inputs.docs).toBeUndefined()
  })
})
