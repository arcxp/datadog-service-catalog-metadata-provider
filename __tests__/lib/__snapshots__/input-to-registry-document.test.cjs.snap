// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`input-to-registry-document.js - schema v2 #inputsToRegistryDocument() - 100% input 1`] = `
{
  "contacts": [
    {
      "contact": "dba-team-name-here@fakeemaildomainthatdoesntexist.com",
      "name": "DBA Team Email Alias",
      "type": "email",
    },
    {
      "contact": "team-name-here@fakeemaildomainthatdoesntexist.com",
      "type": "email",
    },
    {
      "contact": "https://fakeorg.slack.com/archives/A0000000000",
      "type": "slack",
    },
  ],
  "dd-service": "datadog-service-catalog-metadata-provider-test",
  "docs": [
    {
      "name": "GitHub Actions!",
      "provider": "Github",
      "url": "https://github.com/features/actions",
    },
  ],
  "integrations": {
    "opsgenie": {
      "region": "US",
      "service-url": "https://yourorghere.app.opsgenie.com/service/00000000-0000-0000-0000-000000000000",
    },
    "pagerduty": "https://my-org.pagerduty.com/service-directory/PMyService",
  },
  "links": [
    {
      "name": "AMI Version Status Dashboard",
      "type": "dashboard",
      "url": "https://thisisanentirelyfakeurl.seriouslythisisafakehostname.com/dashboard",
    },
  ],
  "repos": [
    {
      "name": "@actions/toolkit",
      "provider": "Github",
      "url": "https://github.com/actions/toolkit",
    },
    {
      "name": "Repo",
      "url": "https://github.com/arcxp/datadog-service-catalog-metadata-provider",
    },
  ],
  "schema-version": "v2",
  "tags": [
    "application:GitHub Action Config Test",
    "env:prod",
    "infrastructure:serverless",
    "language:nodejs",
    "other:value",
  ],
  "team": "Team Name Here",
}
`;

exports[`input-to-registry-document.js - schema v2 #inputsToRegistryDocument() - less input 1`] = `
{
  "contacts": [
    {
      "contact": "team-name-here@fakeemaildomainthatdoesntexist.com",
      "type": "email",
    },
    {
      "contact": "https://fakeorg.slack.com/archives/A0000000000",
      "type": "slack",
    },
  ],
  "dd-service": "datadog-service-catalog-metadata-provider-test",
  "integrations": {
    "opsgenie": {
      "region": "US",
      "service-url": "https://yourorghere.app.opsgenie.com/service/00000000-0000-0000-0000-000000000000",
    },
  },
  "repos": [
    {
      "name": "Repo",
      "url": "https://github.com/arcxp/datadog-service-catalog-metadata-provider",
    },
  ],
  "schema-version": "v2",
  "tags": [
    "application:GitHub Action Config Test",
    "env:prod",
    "infrastructure:serverless",
    "language:nodejs",
  ],
  "team": "Team Name Here",
}
`;

exports[`input-to-registry-document.js - schema v2 #inputsToRegistryDocument() - normal input 1`] = `
{
  "contacts": [
    {
      "contact": "team-name-here@fakeemaildomainthatdoesntexist.com",
      "type": "email",
    },
    {
      "contact": "https://fakeorg.slack.com/archives/A0000000000",
      "type": "slack",
    },
  ],
  "dd-service": "datadog-service-catalog-metadata-provider-test",
  "docs": [
    {
      "name": "GitHub Actions!",
      "provider": "Github",
      "url": "https://github.com/features/actions",
    },
  ],
  "integrations": {
    "opsgenie": {
      "region": "US",
      "service-url": "https://yourorghere.app.opsgenie.com/service/00000000-0000-0000-0000-000000000000",
    },
  },
  "links": [
    {
      "name": "AMI Version Status Dashboard",
      "type": "dashboard",
      "url": "https://thisisanentirelyfakeurl.seriouslythisisafakehostname.com/dashboard",
    },
  ],
  "repos": [
    {
      "name": "@actions/toolkit",
      "provider": "Github",
      "url": "https://github.com/actions/toolkit",
    },
    {
      "name": "Repo",
      "url": "https://github.com/arcxp/datadog-service-catalog-metadata-provider",
    },
  ],
  "schema-version": "v2",
  "tags": [
    "application:GitHub Action Config Test",
    "env:prod",
    "infrastructure:serverless",
    "language:nodejs",
    "other:value",
  ],
  "team": "Team Name Here",
}
`;

exports[`input-to-registry-document.js - schema v2.1 #inputsToRegistryDocument() - 100% input 1`] = `
{
  "contacts": [
    {
      "contact": "dba-team-name-here@fakeemaildomainthatdoesntexist.com",
      "name": "DBA Team Email Alias",
      "type": "email",
    },
    {
      "contact": "team-name-here@fakeemaildomainthatdoesntexist.com",
      "type": "email",
    },
    {
      "contact": "https://fakeorg.slack.com/archives/A0000000000",
      "type": "slack",
    },
  ],
  "dd-service": "datadog-service-catalog-metadata-provider-test",
  "integrations": {
    "opsgenie": {
      "region": "US",
      "service-url": "https://yourorghere.app.opsgenie.com/service/00000000-0000-0000-0000-000000000000",
    },
    "pagerduty": {
      "service-url": "https://my-org.pagerduty.com/service-directory/PMyService",
    },
  },
  "links": [
    {
      "name": "AMI Version Status Dashboard",
      "type": "dashboard",
      "url": "https://thisisanentirelyfakeurl.seriouslythisisafakehostname.com/dashboard",
    },
    {
      "name": "@actions/toolkit",
      "type": "repo",
      "url": "https://github.com/actions/toolkit",
    },
    {
      "name": "GitHub Actions!",
      "type": "doc",
      "url": "https://github.com/features/actions",
    },
    {
      "name": "Repo",
      "type": "repo",
      "url": "https://github.com/arcxp/datadog-service-catalog-metadata-provider",
    },
  ],
  "schema-version": "v2.1",
  "tags": [
    "application:GitHub Action Config Test",
    "env:prod",
    "infrastructure:serverless",
    "language:nodejs",
    "other:value",
  ],
  "team": "Team Name Here",
}
`;

exports[`input-to-registry-document.js - schema v2.1 #inputsToRegistryDocument() - less input 1`] = `
{
  "contacts": [
    {
      "contact": "team-name-here@fakeemaildomainthatdoesntexist.com",
      "type": "email",
    },
    {
      "contact": "https://fakeorg.slack.com/archives/A0000000000",
      "type": "slack",
    },
  ],
  "dd-service": "datadog-service-catalog-metadata-provider-test",
  "integrations": {
    "opsgenie": {
      "region": "US",
      "service-url": "https://yourorghere.app.opsgenie.com/service/00000000-0000-0000-0000-000000000000",
    },
  },
  "links": [
    {
      "name": "Repo",
      "type": "repo",
      "url": "https://github.com/arcxp/datadog-service-catalog-metadata-provider",
    },
  ],
  "schema-version": "v2.1",
  "tags": [
    "application:GitHub Action Config Test",
    "env:prod",
    "infrastructure:serverless",
    "language:nodejs",
  ],
  "team": "Team Name Here",
}
`;

exports[`input-to-registry-document.js - schema v2.1 #inputsToRegistryDocument() - normal input 1`] = `
{
  "contacts": [
    {
      "contact": "team-name-here@fakeemaildomainthatdoesntexist.com",
      "type": "email",
    },
    {
      "contact": "https://fakeorg.slack.com/archives/A0000000000",
      "type": "slack",
    },
  ],
  "dd-service": "datadog-service-catalog-metadata-provider-test",
  "description": "This description describes a service.",
  "integrations": {
    "opsgenie": {
      "region": "US",
      "service-url": "https://yourorghere.app.opsgenie.com/service/00000000-0000-0000-0000-000000000000",
    },
  },
  "lifecycle": "production",
  "links": [
    {
      "name": "@actions/toolkit",
      "provider": "GitHub",
      "type": "repo",
      "url": "https://github.com/actions/toolkit",
    },
    {
      "name": "AMI Version Status Dashboard",
      "type": "dashboard",
      "url": "https://thisisanentirelyfakeurl.seriouslythisisafakehostname.com/dashboard",
    },
    {
      "name": "GitHub Actions!",
      "type": "doc",
      "url": "https://github.com/features/actions",
    },
    {
      "name": "Repo",
      "type": "repo",
      "url": "https://github.com/arcxp/datadog-service-catalog-metadata-provider",
    },
  ],
  "schema-version": "v2.1",
  "tags": [
    "application:GitHub Action Config Test",
    "env:prod",
    "infrastructure:serverless",
    "language:nodejs",
    "other:value",
  ],
  "team": "Team Name Here",
  "tier": "Tier 1",
}
`;
